base/set.jl

```diff
@@ -68,8 +68,17 @@ function union(s::Set, sets::Set...)
     return u
 end
 const ∪ = union
-union!(s::Set, xs) = (for x=xs; push!(s,x); end; s)
-union!(s::Set, xs::AbstractArray) = (sizehint!(s,length(xs));for x=xs; push!(s,x); end; s)
+union!(s::Set, xs) = _union!(s, xs)
+union!(s::Set, xs::AbstractArray) = (sizehint!(s, length(xs)); _union!(s, xs))
```

Comment:
While we're at it, should this maybe dispatch on iteratorsize HasLength/HasSize instead of AbstractArray?